### ISSUE TITLE
Throws error when infobox not found

### ISSUE THREAD

**Issue by michaelsilver** [infolab-csail/WikipediaBase#52]
This is a bug introduced since PR #47. Compare:

Old functionality (03735491463ea3fd95e9885fd982c4edc9e3ad78):
```
>>> from wikipediabase.util import get_infobox, get_meta_infobox
>>> get_meta_infobox('Template:Infobox pan american games saint vincent and the grenadines').rendered_keys()
{}
```
New functionality (2c9ad51fddb5aa55aa115e5e1080d14deda003b7):
```
>>> from wikipediabase.util import get_infobox, get_meta_infobox
>>> get_meta_infobox('Template:Infobox pan american games saint vincent and the grenadines').rendered_keys()
/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/requests-2.7.0-py2.7.egg/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/requests-2.7.0-py2.7.egg/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.linux-x86_64/egg/wikipediabase/util.py", line 89, in get_meta_infobox
  File "build/bdist.linux-x86_64/egg/wikipediabase/util.py", line 175, in _context_get
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox_scraper.py", line 60, in __init__
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox_scraper.py", line 126, in markup_source
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox_scraper.py", line 112, in attributes
  File "build/bdist.linux-x86_64/egg/wikipediabase/article.py", line 79, in html_source
  File "build/bdist.linux-x86_64/egg/wikipediabase/fetcher.py", line 115, in download
  File "build/bdist.linux-x86_64/egg/wikipediabase/fetcher.py", line 97, in _caching_fetch
  File "build/bdist.linux-x86_64/egg/wikipediabase/fetcher.py", line 62, in download
  File "build/bdist.linux-x86_64/egg/wikipediabase/fetcher.py", line 50, in urlopen
LookupError: Error fetching: https://en.wikipedia.org/w/index.php?action=view&redirect=yes&title=Template%3AInfobox+pan+american+games+saint+vincent+and+the+grenadines. Status code 404 : Not Found
```
This is important because the old functionality was quiet and didn't crash programs (such as WikiMap), whereas the new functionality will crash a program, and is not backwards compatible. Is it possible to achieve the good, old quiet functionality when it can't find a template?

**Comment by sfelshin** [comment]
On Aug 22, 2015, at 11:01 PM, Michael Silver <notifications@github.com> wrote:
> This is important because the old functionality was quiet and didn't crash programs (such as WikiMap), whereas the new functionality will crash a program, and is not backwards compatible. Is it possible to achieve the good, old quiet functionality when it can't find a template?

This may be obvious, but if the new functionality is important in some circumstances, someone can add an option saying what to do if the infobox doesn't exist.  In Common Lisp terms, this would look something like

  (defun my-function (...arguments... &key (if-does-not-exist :error))
    (mapc #'(lambda (infobox)
              (cond
               ((...infobox does not exist...)
                (ecase if-does-not-exist
                  (:error ...signal error...)
                  (:ignore nil)
                  ))
               (t ...process infobox...)))
          ...infoboxes...))

-Sue

**Comment by michaelsilver** [comment]
Alvaro says:
> I'll fix this bug; in the meantime, you can wrap the call in a try/except for Wikimap?

Yup, just committed the workaround.

**Comment by michaelsilver** [comment]
Thanks Alvaro for looking into this!

Hmm... I have excepted all `LookupError`'s, but looks like there is another again caused by `get_meta_infobox()`; not sure if this is related or not:
```
Traceback (most recent call last):
  File "/scratch/msilver/.virtualenvs/wikimap/bin/wikimap", line 9, in <module>
    load_entry_point('wikimap==0.1.0', 'console_scripts', 'wikimap')()
  File "/afs/csail.mit.edu/u/m/msilver/wikimap/wikimap/__main__.py", line 23, in main
    args.func(sys.argv[2:])
  File "/afs/csail.mit.edu/u/m/msilver/wikimap/wikimap/create.py", line 49, in main
    G = create_graph(infoboxes, args.clean)
  File "/afs/csail.mit.edu/u/m/msilver/wikimap/wikimap/create.py", line 26, in create_graph
    for unrend, rend in data.get_single_mappings(infobox).iteritems():
  File "/afs/csail.mit.edu/u/m/msilver/wikimap/wikimap/data.py", line 54, in get_single_mappings
    return get_meta_infobox(get_formal_name(infobox)).rendered_keys()
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox_scraper.py", line 139, in rendered_keys
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox.py", line 191, in html_parsed
  File "build/bdist.linux-x86_64/egg/wikipediabase/util.py", line 257, in fromstring
  File "/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/lxml-3.4.4-py2.7-linux-x86_64.egg/lxml/html/__init__.py", line 726, in fromstring
    doc = document_fromstring(html, parser=parser, base_url=base_url, **kw)
  File "/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/lxml-3.4.4-py2.7-linux-x86_64.egg/lxml/html/__init__.py", line 617, in document_fromstring
    "Document is empty")
lxml.etree.ParserError: Document is empty
```
You can ignore everything down to `return get_meta_infobox(get_formal_name(infobox)).rendered_keys()`; that code just calls `get_meta_infobox()` and is highly tested code (I know it's not a problem with _how_ I call `get_meta_infobox()`, but it is a problem with `get_meta_infobox()` itself, I suspect).  Unfortunately, I'm not told what args were passed :(

Thoughs? 
If this is unrelated, I apologize and will create a separate issue.

**Comment by alvaromorales** [comment]
Can you add print statements to see what arguments get passed to get_meta_infobox?

**Comment by michaelsilver** [comment]
Yup, here's the culprit:
```Python
>>> from wikipediabase.util import get_infobox, get_meta_infobox
>>> get_meta_infobox("Template:Infobox comedian awards").rendered_keys()
/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/requests-2.7.0-py2.7.egg/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: A true SSLContext object is not available. This prevents urllib3 from configuring SSL appropriately and may cause certain SSL connections to fail. For more information, see https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox_scraper.py", line 139, in rendered_keys
  File "build/bdist.linux-x86_64/egg/wikipediabase/infobox.py", line 191, in html_parsed
  File "build/bdist.linux-x86_64/egg/wikipediabase/util.py", line 257, in fromstring
  File "/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/lxml-3.4.4-py2.7-linux-x86_64.egg/lxml/html/__init__.py", line 726, in fromstring
    doc = document_fromstring(html, parser=parser, base_url=base_url, **kw)
  File "/scratch/msilver/.virtualenvs/wikimap/local/lib/python2.7/site-packages/lxml-3.4.4-py2.7-linux-x86_64.egg/lxml/html/__init__.py", line 617, in document_fromstring
    "Document is empty")
lxml.etree.ParserError: Document is empty
```
I can put in another except for `lxml.etree.ParserError` in the meantime, but glad I specifically excepted `LookupError` as oppose to just all errors so I caught this one. Yay for error types! :)

**Comment by michaelsilver** [comment]
Ok, so I was able to run `get_meta_infobox()` for every infobox (that I'm aware of) by excepting all `LookupError`'s and `lxml.etree.ParserError`'s, so those two look like the full scope of the issue.

On the bright side of things, once I was able to run my program, I noticed that PR #47 made a noticeable increase (~5%) in the number of returned mappings, decreasing the number of empty or "missing" mappings from before. That's not to mention the great improvement in encoding consistency and data quality. Awesome job, @alvaromorales!
